### PR TITLE
feat: allow tasks to be canceled and rescheduled on different zone in…

### PR DESCRIPTION
**Pull request by mhevery**
… a zone delegate

### CODE REVIEW

**Comment by mhevery** [comment]
TODO
- [ ] Needs a better commit message
- [ ] Add test for task cancelation in ZoneDelegate, which shows how a Task can be rescheduled on a different zone.

**Comment by JiaLiPassion** [comment]
@mhevery, could you explain a little more about the motivation of this PR, I would like to understand more deeply about zone.js, thank you very much.

**Comment by mhevery** [comment]
@JiaLiPassion sorry for the delayed response. 

The issue can best be described by this [test](https://github.com/angular/zone.js/pull/629/files#diff-29a5b9319349c754058e070be5b0215aR235) 

Imagine that you want to write a ZoneSpec which would redirect the scheduling of a task to a different zone. For example you want to make sure that all `addEventListeners` on `scroll` would run outside of some Zone. What you need to do is to intercept the scheduling of the task and than reschedule it on a different zone. However, that could not have been done, since there was no way to tell the `ZoneDelegate`s that the current task, is not part of the current request. We now have `Task. cancelScheduleRequest()` which does just that.

In addition it was hard to keep track of in which state the task was, so I made the Task state explicit, which helps with the error messages.

File: karma-base.conf.js [hunk 1]
```diff
@@ -39,7 +39,7 @@ module.exports = function (config) {
 
     logLevel: config.LOG_INFO,
 
-    browsers: ['Firefox'],
+    browsers: ['ChromeCanary'],
```
Comment (mhevery) [inline karma-base.conf.js]: revert

File: lib/common/timers.ts [hunk 2]
```diff
@@ -71,8 +71,8 @@ export function patchTimer(window: any, setName: string, cancelName: string, nam
       patchMethod(window, cancelName, (delegate: Function) => function(self: any, args: any[]) {
         const task: Task = typeof args[0] === 'number' ? tasksByHandleId[args[0]] : args[0];
         if (task && typeof task.type === 'string') {
-          if (task.cancelFn && task.data.isPeriodic || task.runCount === 0) {
-            // Do not cancel already canceled functions
+          if (task.state !== 'notScheduled' && task.cancelFn ) {
```
Comment (mhevery) [inline lib/common/timers.ts]: revert

File: lib/common/utils.ts [hunk 3]
```diff
@@ -467,9 +467,8 @@ export function createNamedFn(name: string, delegate: (self: any, args: any[]) =
 }
 
 export function patchMethod(
-    target: any, name: string,
-    patchFn: (delegate: Function, delegateName: string, name: string) => (self: any, args: any[]) =>
-        any): Function {
+    target: any, name: string, patchFn: (delegate: Function, delegateName: string, name: string) =>
```
Comment (mhevery) [inline lib/common/utils.ts]: revert

File: lib/zone.ts [hunk 4]
```diff
@@ -229,6 +229,7 @@ interface Zone {
   scheduleEventTask(
       source: string, callback: Function, data: TaskData, customSchedule: (task: Task) => void,
       customCancel: (task: Task) => void): EventTask;
+  scheduleTask<T extends Task>(task: T): T;
```
Comment (mhevery) [inline lib/zone.ts]: add docs

File: lib/zone.ts [hunk 5]
```diff
@@ -516,6 +527,8 @@ interface Task {
    * Number of times the task has been executed, or -1 if canceled.
    */
   runCount: number;
+
+  cancelScheduleRequest(): void;
```
Comment (mhevery) [inline lib/zone.ts]: add docs

File: lib/zone.ts [hunk 6]
```diff
@@ -622,13 +639,13 @@ const Zone: ZoneType = (function(global: any) {
     }
 
     public fork(zoneSpec: ZoneSpec): AmbientZone {
-      if (!zoneSpec) throw new Error('ZoneSpec required!');
+      if (!zoneSpec) throw Error('ZoneSpec required!');
```
Comment (mhevery) [inline lib/zone.ts]: revert

**Comment by JiaLiPassion** [comment]
@mhevery , Thank you for explaining, I will read the source and learn how it works!

**Pull request merged**